File: archive/logic/TuringMachineSimulationNew/Heads/IHead.cs
using TuringMachineSimulation.Logic.Tapes;

namespace TuringMachineSimulation.Logic.Heads
{
    public interface IHead
    {
        // Allow UIs to know which head it is and render it correctly.
        public HeadType Type { get; }

        // Leave this public for UI to get contents near the head including left and right.
        public int Position { get; internal set; }
        
        // Leave this public for UI to get the referencing tape
        public int TapeID { get; internal set; }

        /// <summary> Gets the content in current position of the head. </summary>
        /// <returns> Returns the read content. Can be null when out of bound. </returns>
        public char? GetCurrentContent();
        
        /// <summary> Moves the head. No effects to the tape. </summary>
        /// <param name="steps"> Moves this number of steps. Can be negative and zero. </param>
        public void Move(int steps);

        /// <summary> Try to write to content into a cell on the tape </summary>
        /// <param name="content"> The content that is going to store in the cell. </param>
        /// <param name="machineID"> The machine ID. </param>
        /// <param name="headID"> The head ID. </param>
        /// <returns> True if the write operation is successful. False otherwise. </returns>
        public bool TryWrite(char content, int machineID, int headID);

        /// <summary> Check if the head is using the given tape. </summary>
        /// <param name="tape"> The given ITape object. </param>
        /// <returns> True if it is using the given tape. </returns>
        public bool IsUsesTape(ITape tape);
    }
}

File: frontend/turing-machine-web/src/logic/Heads/IHead.ts
import { HeadTypes } from "./HeadTypes"
import { ITape } from "../Tapes/ITape"
import { SignalState } from "../States/SignalStates";

export interface IHead 
{
    readonly Type: HeadTypes;
    Position: number;
    TapeID: number;

    /**
     * Gets the content in the current position of the head.
     * @returns Returns the read content. Can be null when out of bound.
     */
    GetCurrentContent(): string | null;

    /**
     * Receives the control signal in the current position from the tape.
     * @returns Returns the control signal. Receive Other if no state is written.
     */
    ReceiveSignal(): SignalState;

    /**
     * Moves the head. No effects to the tape.
     * @param steps: Moves this number of steps. Can be negative or zero.
     */
    Move(steps: number): void;

    /**
     * Tries to write a content into a cell on the tape
     * @param content The write content. It must be a single character.
     * @param machineID The machine that is going to write.
     * @param headID The machine head that is going to write.
     * @returns A boolean value indicating success or not.
     */
    TryWrite(content: string, machineID: number, headID: number): boolean;

    /**
     * Removes the signal from the tape.
     */
    TakeOutSignal(): void;

    /**
     * Checks if the head is using the given tape.
     * @returns True if the given ITape object is the same as the one this head is using, False otherwise.
     */
    IsUsesTape(tape: ITape): boolean;
}